### DISCUSSION TITLE
GraphSage unsup implementation in ./examples

### DISCUSSION THREAD

**Discussion by OliverRock** [pyg-team/pytorch_geometric#8683] [General]
I am trying to implement an unsupervised GraphSage model but I am slightly confused by [this example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup.py#L49). I think the loss function here is not the same as in the original [paper](https://arxiv.org/abs/1706.02216) because the code does not use random walk to find the nodes and it is passing the labels so it is not unsupervised. Are these observations correct or have I misunderstood something?

Here is the part of the paper which defines the loss function:

![JPEG image-499F-8C3E-58-0](https://github.com/pyg-team/pytorch_geometric/assets/32681524/c524341b-ce7f-46ca-9f3d-4be8199e375c)

**Comment by rusty1s** [comment]
Yes, the `GraphSAGE` unsupervised example doesn't use fixed-length random walks as negative samples but randomly sampled one. It is still unsupervised (not sure what you mean by it is using the label); it uses positive and negative links to compute the constrastive loss.
